toy-problems/islands/islands.js

```diff
@@ -0,0 +1,83 @@
+/*
+two a dimensial grid map
+each element is going to represent a peice of land
+1 is land, 0 is water
+ output a number which is the number of islands
+
+Example 1:
+  Input:
+  11110
+  11010
+  11000
+  00000
+
+  Output: 1
+
+Example 2:
+  Input:
+  11000
+  11000
+  00100
+  00011
+
+  Output: 3
+
+  I: two dimensional array
+  O: a single integer; total number of islands
+
+Pseudocode:
+  OUTER FUNCTION
+    set count to 0
+
+    INNER FUNCTION - flood (col, row)
+      if the tile is water
+        return
+      make tile water(flood tile)
+      invoke flood on the neighbor coordinates
+
+    iterate over the matrix (col, row)
+      if the current element is a 1
+        increment count
+        invoke flood (coordinates for col and row)
+
+    Return the count
+*/
+const grid = [
+  ['1', '1', '0', '0', '0'],
+  ['1', '1', '0', '0', '0'],
+  ['0', '0', '1', '0', '0'],
+  ['0', '0', '0', '1', '1'],
+];
+
+const islands = (matrixGrid) => {
+  const matrix = matrixGrid;
```

Comment:
These types normally allow (and leetcode style problems seems to prefer) manipulation of the arg directly (I know airbnb style guide hates this).